content/blog/withr-3-0-0/index.Rmd

```diff
@@ -0,0 +1,226 @@
+---
+output: hugodown::hugo_document
+
+slug: withr-3-0-0
+title: withr 3.0.0
+date: 2024-01-17
+author: Lionel Henry
+description: >
+    withr is the tidyverse solution for automatically cleaning
+    up after yourselves (temporary files, options, etc). This milestone makes withr much faster.
+
+photo:
+  url: https://unsplash.com/photos/brown-and-black-brush-on-brown-wooden-table-V0cSTljC92k
+  author: Neal E. Johnson
+
+categories: [package]
+tags: [r-lib, withr]
+---
+
+It's not without jubilant bearing that we announce the release of the 3.0.0 version of [withr](https://withr.r-lib.org/), the tidyverse solution for automatic cleanup of resources! In this release, the internals of withr were rewritten to improve the performance and increase the compatibility with base R's `on.exit()` mechanism.
+
+You can install it from CRAN with:
+
+```{r, eval = FALSE}
+install.packages("withr")
+```
+
+In this blog post we'll go over the changes that made this rewrite possible, but first we'll rewiew the cleanup strategies made possible by withr.
+
+You can see a full list of changes in the [release notes](https://withr.r-lib.org/news/index.html#withr-300).
+
+```{r setup, echo = FALSE}
+# To avoid distracting calls to `eval()` in `local()`
+stop <- rlang::abort
+```
+
+
+## Cleaning up resources with base R and with withr
+
+Traditionally, resource cleanups in R is done with `base::on.exit()`. Cleaning up in the on-exit hook ensures that the cleanup happens both in the normal case, when the code has finished running without error, and in the error case, when something went wrong and execution is interrupted.
+
+`on.exit()` is meant to be used inside functions but it also works within `local()`, which we'll use here for our examples:
+
+```{r}
+local({
+  on.exit(message("Cleaning time!"))
+  1 + 2
+})
+```
```

Comment:
I was actually surprised to see "Cleaning time!" printed *before* "3". And this is especially surprising given an example below. I guess it had to do with automatic printing?

In any case, maybe it's better to do 2 explicit things here and see them unfold in the expected order.

Reply:
Good point